File: frontend/src/services/expenseService.js
import API from "./api";

export const getExpenses =
  async (
    search = "",
    page = 0
  ) => {

    const response =
      await API.get(

        `/expenses?search=${search}&page=${page}&size=10`

      );

    return response.data;

};

export const addExpense =
  async (data) => {

    const response =
      await API.post(
        "/expenses",
        data
      );

    return response.data;

};

export const deleteExpense =
  async (id) => {

    const response =
      await API.delete(
        `/expenses/${id}`
      );

    return response.data;

};
export const updateExpense =
  async (
    id,
    data
  ) => {

    const response =
      await API.put(

        `/expenses/${id}`,
        data

      );

    return response.data;

};

File: backend/AI-Powered-Expense-Tracker/src/main/java/org/expense/service/ExpenseService.java
package org.expense.service;

import java.time.LocalDate;
import java.util.List;

import org.expense.dto.ExpenseRequest;
import org.expense.entity.Expense;
import org.springframework.data.domain.Page;

public interface ExpenseService {

    Expense addExpense(ExpenseRequest request);

    List<Expense> getAllExpenses();
    
    Expense updateExpense(Long id, ExpenseRequest request);

    String deleteExpense(Long id);
    Page<Expense> getExpenses(
            int page,
            int size,
            String sortBy,
            String direction);
    
    List<Expense> searchExpenses(String keyword);
    
    Page<Expense> filterExpenses(
            int page,
            int size,
            String search,
            String sortBy,
            String direction,
            Double minAmount,
            Double maxAmount,
            LocalDate startDate,
            LocalDate endDate);

}